components/net/http_cache.rs

```diff
@@ -0,0 +1,488 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#![deny(missing_docs)]
+
+//! A memory cache Implements the logic specified in http://tools.ietf.org/html/rfc7234
+//! and http://tools.ietf.org/html/rfc7232.
+
+use fetch::methods::DoneChannel;
+use hyper::header;
+use hyper::header::ContentType;
+use hyper::header::Headers;
+use hyper::method::Method;
+use hyper::status::StatusCode;
+use hyper_serde::Serde;
+use net_traits::{Metadata, FetchMetadata, FilteredMetadata};
+use net_traits::request::Request;
+use net_traits::response::{Response, ResponseBody};
+use servo_url::ServoUrl;
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::{Arc, Mutex};
+use time;
+use time::{Duration, Tm};
+use url::Url;
+
+
+/// The key used to differentiate requests in the cache.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct CacheKey {
+    url: ServoUrl
+}
+
+impl CacheKey {
+    fn new(request: Request) -> CacheKey {
+        CacheKey {
+            url: request.url().clone()
+        }
+    }
+
+    fn from_url_string(url: String) -> CacheKey {
+        let url = Url::parse(&url).unwrap();
+        CacheKey {
+            url: ServoUrl::from_url(url)
+        }
+    }
+
+    /// Retrieve the URL associated with this key
+    pub fn url(&self) -> ServoUrl {
+        self.url.clone()
+    }
+}
+
+/// A complete cached resource.
+#[derive(Clone, Debug)]
+struct CachedResource {
+    metadata: CachedMetadata,
+    request_headers: Vec<(String, String)>,
+    body: Arc<Mutex<ResponseBody>>,
+    status: Option<StatusCode>,
+    raw_status: Option<(u16, Vec<u8>)>,
+    url_list: Vec<ServoUrl>,
+    expires: Arc<Mutex<Duration>>,
+    last_validated: Tm
+}
+
+/// Metadata about a loaded resource, such as is obtained from HTTP headers.
+#[derive(Clone, Debug)]
+struct CachedMetadata {
+    /// Final URL after redirects.
+    pub final_url: ServoUrl,
+    /// MIME type / subtype.
+    pub content_type: Option<Serde<ContentType>>,
+    /// Character set.
+    pub charset: Option<String>,
+    /// Headers
+    pub headers: Arc<Mutex<Headers>>,
```

Comment:
Why do the headers need their own mutex?

Reply:
I did that because some internal part of hyper::Headers didn't implement the right traits to make it shareable across threads.

Reply:
Why do you need to share the headers across threads? Aren't you supposed to be cloning them instead?

Reply:
The problem seems to occur here(and a few other places where a thread is build) https://github.com/gterzian/servo/blob/ab5c9798a03cdcf24704fd418cf9db827e3d3514/components/net/resource_thread.rs#L337

```
error[E0277]: the trait bound `std::cell::UnsafeCell<serde::export::Option<std::vec::Vec<std::vec::Vec<u8>>>>: std::marker::Sync` is not satisfied in `(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)`
   --> /Users/gregory/Projects/servo/components/net/resource_thread.rs:337:83
    |
337 |         thread::Builder::new().name(format!("fetch thread for {}", req_init.url)).spawn(move || {
    |                                                                                   ^^^^^ `std::cell::UnsafeCell<serde::export::Option<std::vec::Vec<std::vec::Vec<u8>>>>` cannot be shared between threads safely
    |
    = help: within `(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)`, the trait `std::marker::Sync` is not implemented for `std::cell::UnsafeCell<serde::export::Option<std::vec::Vec<std::vec::Vec<u8>>>>`
    = note: required because it appears within the type `hyper::header::internals::cell::OptCell<std::vec::Vec<std::vec::Vec<u8>>>`
    = note: required because it appears within the type `hyper::header::internals::item::Item`
    = note: required because it appears within the type `(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)`
    = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)>`
    = note: required because it appears within the type `alloc::raw_vec::RawVec<(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)>`
    = note: required because it appears within the type `std::vec::Vec<(unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item)>`
    = note: required because it appears within the type `hyper::header::internals::vec_map::VecMap<unicase::UniCase<hyper::header::CowStr>, hyper::header::internals::item::Item>`
    = note: required because it appears within the type `hyper::header::Headers`
    = note: required because it appears within the type `http_cache::CachedMetadata`
    = note: required because it appears within the type `http_cache::CachedResource`
    = note: required because of the requirements on the impl of `std::marker::Sync` for `std::ptr::Unique<http_cache::CachedResource>`
    = note: required because it appears within the type `alloc::raw_vec::RawVec<http_cache::CachedResource>`
    = note: required because it appears within the type `std::vec::Vec<http_cache::CachedResource>`
    = note: required because of the requirements on the impl of `std::marker::Sync` for `std::collections::hash::table::RawTable<http_cache::CacheKey, std::vec::Vec<http_cache::CachedResource>>`
    = note: required because it appears within the type `std::collections::HashMap<http_cache::CacheKey, std::vec::Vec<http_cache::CachedResource>>`
    = note: required because it appears within the type `http_cache::HttpCache`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::RwLock<http_cache::HttpCache>`
    = note: required because it appears within the type `http_loader::HttpState`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<http_loader::HttpState>`
    = note: required because it appears within the type `[closure@/Users/gregory/Projects/servo/components/net/resource_thread.rs:337:89: 363:10 req_init:net_traits::request::RequestInit, http_state:std::sync::Arc<http_loader::HttpState>, ua:std::borrow::Cow<'_, str>, dc:serde::export::Option<std::sync::mpsc::Sender<devtools_traits::DevtoolsControlMsg>>, filemanager:filemanager_thread::FileManager, res_init_:serde::export::Option<net_traits::response::ResponseInit>, sender:ipc_channel::ipc::IpcSender<net_traits::FetchResponseMsg>]`
```